### PR TITLE
poc: Generate test vectors for IdpfPoplar

### PR DESCRIPTION
**Pull request by cjpatton**
For a given input `alpha`, generate the expected public and key shares.

### CODE REVIEW

**Comment by cjpatton** [comment]
cc/ @divergentdave this might help you get going. Note that the public share looks highly non-random because the IPDF key shares are identical:
```
{
    "beta_inner": [
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ],
        [
            1,
            1
        ]
    ],
    "beta_leaf": [
        1,
        1
    ],
    "keys": [
        "01010101010101010101010101010101",
        "01010101010101010101010101010101"
    ],
    "public_share": "055555000000000000000000000000000000000000000000000001000000000000000100000000000000000000000000000000ffffffff00000000ffffffff0000000000000000000000000000000000000000ffffffff00000000ffffffff0000000000000000000000000000000000000000ffffffff00000000ffffffff0000000000000000000000000000000000000000ffffffff00000000ffffffff000000000000000000000000000000000000000000000000000000010000000000000001000000000000000000000000000000000000000000000001000000000000000100000000000000000000000000000000000000000000000100000000000000010000000000000000000000000000000000000000000000010000000000000001000000000000000000000000000000007fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffec7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffec"
}
```

File: poc/idpf.sage [hunk 1]
```diff
@@ -117,6 +119,31 @@ def test_idpf(Idpf, alpha, level, prefixes):
...
 
+def gen_test_vec(Idpf, alpha, test_vec_instance):
+    beta_inner = [[Idpf.FieldInner(1)] * Idpf.VALUE_LEN] * (Idpf.BITS-1)
+    beta_leaf = [Idpf.FieldLeaf(1)] * Idpf.VALUE_LEN
+    (public_share, keys) = Idpf.gen(alpha, beta_inner, beta_leaf)
+
+    printable_beta_inner = [
+        [ elem.as_unsigned() for elem in value ] for value in beta_inner
+    ]
+    printable_beta_leaf = [ elem.as_unsigned() for elem in beta_leaf ]
```
Comment (divergentdave) [inline poc/idpf.sage:130]: These should be stringified to avoid JSON library compatibility issues

File: poc/idpf.sage [hunk 2]
```diff
@@ -117,6 +119,31 @@ def test_idpf(Idpf, alpha, level, prefixes):
...
+    printable_keys = [ key.hex() for key in keys ]
+    test_vec = {
+        'beta_inner': printable_beta_inner,
+        'beta_leaf': printable_beta_leaf,
+        'public_share': public_share.hex(),
+        'keys': printable_keys,
+    }
```
Comment (divergentdave) [inline poc/idpf.sage:137]: The test vector should include alpha as well, so that it's self-contained. (Perhaps as an array of `0` and `1`?)

Comment (cjpatton) [reply poc/idpf.sage:137]: Are you alright with an unsinged integer here? From the spec perspective, this is the most natural way to represent `alpha`.

Comment (divergentdave) [reply poc/idpf.sage:137]: If we included `bits` as a number and `alpha` as a number-as-a-string, that would work.

File: poc/idpf.sage (hunk 1 above)
Comment (cjpatton) [reply poc/idpf.sage]: Done.

File: poc/idpf.sage (hunk 2 above)
Comment (cjpatton) [reply poc/idpf.sage]: Done.

File: poc/idpf.sage [hunk 3]
```diff
@@ -117,6 +119,32 @@ def test_idpf(Idpf, alpha, level, prefixes):
...
+    (public_share, keys) = Idpf.gen(alpha, beta_inner, beta_leaf)
+
+    printable_beta_inner = [
+        [ str(elem.as_unsigned()) for elem in value ] for value in beta_inner
+    ]
+    printable_beta_leaf = [ str(elem.as_unsigned()) for elem in beta_leaf ]
+    printable_keys = [ key.hex() for key in keys ]
+    test_vec = {
+        'bits': int(Idpf.BITS),
+        'beta_inner': printable_beta_inner,
```
Comment (divergentdave) [inline poc/idpf.sage:134]:
```suggestion
        'bits': int(Idpf.BITS),
        'alpha': str(alpha),
        'beta_inner': printable_beta_inner,
```

File: poc/idpf.sage [hunk 4]
```diff
@@ -117,6 +119,32 @@ def test_idpf(Idpf, alpha, level, prefixes):
...
+        'beta_leaf': printable_beta_leaf,
+        'public_share': public_share.hex(),
+        'keys': printable_keys,
+    }
+
+    os.system('mkdir -p test_vec/{}'.format(DRAFT))
+    with open('test_vec/{}/{}_{}.json'.format(
+        DRAFT, Idpf.__name__, test_vec_instance), 'w') as f:
```
Comment (divergentdave) [inline poc/idpf.sage:142]: We can punt on this for now, but we'll soon need to include the PRG name here.

**Review by divergentdave — CHANGES_REQUESTED** [review]

File: poc/idpf.sage (hunk 4 above)
Comment (cjpatton) [reply poc/idpf.sage:142]: Done.

**Comment by cjpatton** [comment]
@divergentdave just a heads up that I update the value of `alpha` used for unit tests so that bit order is easier to test.

**Review by divergentdave — APPROVED** [review]

**Comment by cjpatton** [comment]
- [ ] To make this more useful, the key shares should be distinct strings so that the correction words don't end up canceling out.

**Comment by cjpatton** [comment]
Closed in favor of #162.

**Pull request closed without merging**